internals/overlord/planstate/manager.go

```diff
@@ -76,14 +74,16 @@ type PlanChangedFunc func(p *plan.Plan)
 // change event does not guarantee that combined plan content has changed.
 // Notification registration must be completed before the plan is loaded.
 func (m *PlanManager) AddChangeListener(f PlanChangedFunc) {
-	m.planLock.Lock()
-	defer m.planLock.Unlock()
-	m.planHandlers = append(m.planHandlers, f)
+	m.changeListeners = append(m.changeListeners, f)
 }
 
-func (m *PlanManager) planChanged(plan *plan.Plan) {
-	m.plan = plan
-	for _, f := range m.planHandlers {
+func (m *PlanManager) callChangeListeners(plan *plan.Plan) {
+	if plan == nil {
+		// Avoids if statement on every deferred call to this method (we
+		// shouldn't call listeners when the operation fails).
+		return
+	}
```

Comment:
If we also had a `recover()` higher up, we'd want to recover here then re-panic rather than call the changeListeners. But I don't think we recover anywhere do we?

Reply:
We do recover and re-panic in Run(). However, in case of a panic in the functions that call `callChangeListeners`, plan (newPlan) will be nil, so it won't call them due to this `if` statement.